### PR TITLE
[DI] Overriding services autowired by name under _defaults bind not working

### PR DESCRIPTION
**Pull request by przemyslaw-bogusz**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #28326 
| License       | MIT

This is an implementation of ideas and suggestions of @nicolas-grekas and @GuilhemN.

### CODE REVIEW

**Comment by OskarStark** [comment]
Typo in PR headline:
```diff
- [DI] Fixes: #28326 - Overriding services autowired by name under _deaults bind not working
+ [DI] Fixes: #28326 - Overriding services autowired by name under _defaults bind not working
```

**Review by GuilhemN — COMMENTED** [review]
This is great :+1: I'm only concerned about the `PhpFileLoader` which doesn't call `FileDefinition::setDefinition()` (https://github.com/przemyslaw-bogusz/symfony/blob/85a28f1e4e5a6197bbc4ed0550b9ce68936c83bf/src/Symfony/Component/DependencyInjection/Loader/Configurator/ServiceConfigurator.php#L65).
Shouldn't the logic be added there too?

**Comment by nicolas-grekas** [comment]
See comment in #29949: there may be an issue that makes the container load all classes at compile time. We should avoid it if confirmed.

**Comment by przemyslaw-bogusz** [comment]
@GuilhemN Thanks for the suggestion. I added the funcionality for `PhpFileLoader`, and as a consequence I moved code from `FileLoader` into `ContainerBuilder` to avoid duplication. 

@nicolas-grekas This PR is rather simple so I don't think I've made a change that would cause 

> the container load all classes at compile time

At least, I'm not aware of that.

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php [hunk 1]
```diff
@@ -1504,6 +1506,35 @@ public function normalizeId($id)
         return isset($this->definitions[$id]) || isset($this->aliasDefinitions[$id]) || isset($this->removedIds[$id]) ? $id : parent::normalizeId($id);
     }
 
+    /**
+     * Gets removed binding ids.
+     *
+     * @return array
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: please keep only `@internal` in the docblock (no title) and use a real return type

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php [hunk 2]
```diff
@@ -1504,6 +1506,35 @@ public function normalizeId($id)
...
+    }
+
+    /**
+     * Adds a removed binding id.
+     *
+     * @param int $id
+     *
+     * @internal
+     */
+    public function addRemovedBindingIds($id)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: `int` instead of annotation+docblock title

**Review by nicolas-grekas — COMMENTED** [review]
cannot test right now,here is a very light review

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php (hunk 1 above)
Comment (OskarStark) [reply src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: Are you sure with the return type, it’s against 3.4 branch, or am I mixing things up?

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: oh, 3.4, my bad!

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @przemyslaw-bogusz.

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php (hunk 2 above)
Comment (teohhanhui) [reply src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: This should actually be `string`.

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: good catch, could you please send a PR to fix it?

Comment (teohhanhui) [reply src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: https://github.com/symfony/symfony/pull/31173

**Pull request merged**